### ISSUE TITLE
onconnect on first run

### ISSUE THREAD

**Issue by marv-th** [Splamy/TS3AudioBot#606]
**Describe the bug**
When you run the bot for the first time and before this first run there is an ts3audiobot.toml and an onconnect event like that (onconnect = "!xecute (!plugin load 2)"), than it does not work. 
Only on the second run.

**Expected behavior**
I want to load plugins when the bot is starting for the first time.

**To Reproduce**
Start the bot a second time

**Version**
Version: 0.10.0-alpha.144
Branch: develop
CommitHash: 03b03ed8fc68f0c7df601b11ca7068a2390a6643

**Platform**
Debian 9 with dotnet

There is no log, because, they dont printing no errors for that or anywhere.

Hope you can find a possible resolution for me.

**Comment by Splamy** [comment]
```js
"Splamy": !setting set events.onconnect "!plug loa Example2.cs"
(restart)
"Splamy": !plug lis
"TS3AudioBot": All available plugins: 
...
#02|+ON|Example2 (BotPlugin)
```

I've tried that but unfortunately I can't reproduce that problem. Try loading the plugin via name instead of id. Id's might change when you restart or add/remove files to the plugins folder.

**Comment by marv-th** [comment]
I think i found my problem. Is it not otherwice possible to load the plugin before one bot is connect? like an onstart event?

**Comment by Splamy** [comment]
I assume with 'onstart' you want something like a 'beforeconnect' event?
Or rather, what is your use case /  Can you explain what do you want to archive?
This would help me get an understanding what your requirements are.

There are multiple stages at where I could load plugins, and for what I can make config events.
Eg, the order of stages is something like:
- `beforeinitializing`: when an new Bot class gets created
- `earliestinitialized`: when all modules the plugin depends on are loaded (ths is the same as `beforeinitializing` if the plugin has no dependencies)
- `afterinitialized`, aka `beforeconnect`: after all modules are loaded
- `onconnect` aka `afterconnect`: after the bot connected successfully

Here is a table of what is possible and what is considered

|                       | Plugin        | Event            |
|-----------------------|---------------|------------------|
| `beforeinitializing`  | 🟡 problematic | ❌ not possible   |
| `earliestinitialized` | ✔️ possible    | ❌ not possible   |
| `beforeconnect`       | ✔️ possible    | 🟡 problematic    |
| `onconnect`           | 🟡 possible, but not useful    | ✔️ Already exists |

The current `onconnect = !plugin load ...` is currently the only possible way to load plugins on start.
However the [plugin.load] will have more options how I want to implement it. I think `earliestinitialized` will probably be a good choice, but it's more complex to implement than `beforeconnect`, and could be more error prone, and more complex to understand.

**Comment by marv-th** [comment]
I will post my answer in german because i can not explain it correctly in Englisch. Sorry for that :(

Also es ist so, dass wir die Installationsweise eines einzelnen Bots anders durchführen. Deshalb haben wir ein Plugin, welches ein Script (welches eine datei im bots ordner erstellt, in welcher die vom user angepassten Einstellungen gespeichert werden) ausführt, sobald ein benutzerdefinierter Befehl ausgeführt wurde (chat/api). Der Ts3audiobot wird durch ein selbst geschriebenes Installations Script so installiert, dass kein Bot erstellt ist oder erstellt werden muss. Damit der User bots im Webpanel erstellen kann muss halt das plugin geladen sein, welches aber, soweit ich es verstehe, nicht geladen wird, wenn kein bot erstellt oder aktiv ist. Der Grund dafür ist ja das onconnect event.

**Comment by Splamy** [comment]
So, not a bot plugin but a core/static plugin.
Currently there's no way, that would be what the `[plugins]` setting is planned for.

**Comment by marv-th** [comment]
Okay! Im looking forward to that!😃

**Comment by Splamy** [comment]
Closed in favor #650
